airflow/jobs/scheduler_job.py

```diff
@@ -901,6 +901,33 @@ def __get_concurrency_maps(
             task_map[(dag_id, task_id)] = count
         return dag_map, task_map
 
+    def _remove_tis_with_missing_dag(self, task_instances, session=None):
+        """
+        Fail task instances and the corresponding DagRun if the dag can't be found in
+        the dags folder but exists in SerializedDag table.
+        Return task instances that exists in SerializedDag table as well as dags folder.
+        If the dag can't be found in DagBag, just return the task instance. This is common in
+        unittest where subdir is os.devnull
+        """
+        tis = []
+        for ti in task_instances:
+            try:
+                dag = self.dagbag.get_dag(ti.dag_id, session=session)
+            except SerializedDagNotFound:
+                tis.append(ti)
+                continue
+            if os.path.exists(dag.fileloc):
+                tis.append(ti)
+            else:
+                dagrun = dag.get_dagrun(execution_date=ti.execution_date, session=session)
```

Comment:
No need to use the dag object here.

```suggestion
                dagrun = ti.get_dagrun(session)
```